codalab/server/bundle_manager.py

```diff
@@ -299,16 +299,25 @@ def _bring_offline_stuck_running_bundles(self, workers):
                 logger.info('Bringing bundle offline %s: %s', bundle.uuid, failure_message)
                 self._model.transition_bundle_worker_offline(bundle)
 
-    def _schedule_run_bundles_on_workers(self, workers, user_owned):
+    def _schedule_run_bundles_on_workers(self, workers, staged_bundles_to_run):
         """
         Schedules STAGED bundles to run on the given workers. If user_owned is
         True, then schedules on workers run by the owner of each bundle.
         Otherwise, uses CodaLab-owned workers, which have user ID root_user_id.
         """
-        for bundle in self._model.batch_get_bundles(state=State.STAGED, bundle_type='run'):
-            if user_owned:
-                workers_list = workers.user_owned_workers(bundle.owner_id)
-            else:
+        # Reorder the stage_bundles so that bundles which were requested to run on a personal worker
+        # will be scheduled to run first
+        staged_bundles_to_run.sort(
+            key=lambda b: (b.metadata.request_queue is not None, b.metadata.request_queue),
+            reverse=True,
+        )
+
+        for bundle, bundle_resources in staged_bundles_to_run:
+            # Check if there is an user_owned worker.
+            # If so, schedule the current bundle to run on a user_owned worker first
+            workers_list = workers.user_owned_workers(bundle.owner_id)
```

Comment:
I think this is a good point! We should always be better at utilizing all the current resources we have and unblocking users' jobs. I have updated the code to fall back to public workers so that even if there is no valid private workers available, we can still check our public workers to see if the current bundle can be run there.